### PR TITLE
Support of Hadoop archive in HdfsArtifactRepository

### PR DESCRIPTION
**Pull request by jdlesage**
## What changes are proposed in this pull request?
 
A run artifacts can now be stored in a Hadoop Archive (https://hadoop.apache.org/docs/r1.2.1/hadoop_archives.html). This PR  implements the support of these archive in order to browse and download artifacts

A next PR will implement the command to pack all the artifacts of a run into a hadoop archive. It will enable to reduce number of files on HDFS (number of files have a big impact on HDFS performance)
 
## How is this patch tested?
 
* Unit tests
* Running the server with a hadoop archive as artifact
 
### Is this a user-facing change? 

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [ ] Tracking
- [ ] Projects 
- [X] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by jdlesage** [comment]
An example of usage:

image

File: mlflow/store/hdfs_artifact_repo.py [hunk 1]
```diff
@@ -159,7 +163,12 @@ def hdfs_system(host, port):
     """
     import pyarrow as pa
 
-    driver = os.getenv('MLFLOW_HDFS_DRIVER') or 'libhdfs'
+    if host and host.startswith("har://"):
+        _logger.info(
+            "Hadoop archive can be open only by libhdfs. Override MLFLOW_HDFS_DRIVER setting.")
```
Comment (vguerra) [inline mlflow/store/hdfs_artifact_repo.py]: A more accurate message would probably be ".... Ignoring MLFLOW_HDFS_DRIVER setting.

```suggestion
            "Hadoop archive can be open only by libhdfs. Ignoring MLFLOW_HDFS_DRIVER setting.")
```

File: mlflow/store/hdfs_artifact_repo.py [hunk 2]
```diff
@@ -174,11 +183,27 @@ def hdfs_system(host, port):
...
 def _resolve_connection_params(artifact_uri):
     parsed = urllib.parse.urlparse(artifact_uri)
+    if parsed.scheme == "har":
+        return _parse_har_filesystem(artifact_uri)
     return parsed.hostname, parsed.port, parsed.path
 
 
+def _parse_har_filesystem(artifact_uri):
+    har_extension_index = artifact_uri.find(HAR_EXTENSION)
+    if har_extension_index < 0:
```
Comment (vguerra) [inline mlflow/store/hdfs_artifact_repo.py]: Wouldn't something like `artifact_uri.endswith(HAR_EXTENSION)` be more robust? .. with this approach I guess that having HAR_EXTENSION appearing even in the middle of the `artifact_uri` would pass this check

File: tests/store/test_hdfs_artifact_repo.py [hunk 3]
```diff
@@ -44,7 +46,7 @@ def test_log_artifact_with_kerberos_setup(hdfs_system_mock):
     os.environ['MLFLOW_KERBEROS_USER'] = 'some_kerberos_user'
     os.environ['MLFLOW_HDFS_DRIVER'] = 'libhdfs3'
 
-    repo = HdfsArtifactRepository('hdfs:/some/maybe/path')
+    repo = HdfsArtifactRepository('hdfs:///some/maybe/path')
```
Comment (vguerra) [inline tests/store/test_hdfs_artifact_repo.py]: why 3 slashes?

File: mlflow/store/hdfs_artifact_repo.py (hunk 2 above)
Comment (jdlesage) [reply mlflow/store/hdfs_artifact_repo.py]: This function can use a path inside a archive as artifact folder. So .har will be in the middle of the `artifact_uri`

So it is an expected behavior

File: tests/store/test_hdfs_artifact_repo.py (hunk 3 above)
Comment (jdlesage) [reply tests/store/test_hdfs_artifact_repo.py]: The format is `{schema}://{host}/{path}` if `schema = hdfs` and `host = root` it becomes `hdfs://root/user` 
`root` can be omitted, it becomes `hdfs:///user`

**Pull request merged**